src/Agent.Worker/DiagnosticLogManager.cs

```diff
@@ -0,0 +1,262 @@
+            using (StreamWriter writer = File.CreateText(environmentFile)) 
+            {
+                writer.Write(content);
+            }
+
+            // Copy worker diag log files
+            List<string> workerDiagLogFiles = GetWorkerDiagLogFiles(HostContext.GetDirectory(WellKnownDirectory.Diag), jobStartTimeUtc);
+            executionContext.Debug($"Copying {workerDiagLogFiles.Count()} worker diag logs.");
+
+            foreach(string workerLogFile in workerDiagLogFiles)
+            {
+                ArgUtil.Directory(workerLogFile, nameof(workerLogFile));
+
+                string destination = Path.Combine(supportFilesFolder, Path.GetFileName(workerLogFile));
+                File.Copy(workerLogFile, destination);
+            }
+            
+            executionContext.Debug("Zipping diagnostic files.");
+
+            string buildNumber;
+            if (!message.Environment.Variables.TryGetValue(Constants.Variables.Build.Number, out buildNumber))
+            {
+                buildNumber = "UnknownBuildNumber";
+            }
+            string buildName = $"Build {buildNumber}";
+            
+            string phaseName;
+            if (!message.Environment.Variables.TryGetValue(Constants.Variables.System.PhaseDisplayName, out phaseName))
+            {
+                phaseName = "UnknownPhaseName";
+            }
+
+            // zip the files
+            string diagnosticsZipFileName = $"{buildName}-{phaseName}.zip";
+            string diagnosticsZipFilePath = Path.Combine(supportRootFolder, diagnosticsZipFileName);
+            ZipFile.CreateFromDirectory(supportFilesFolder, diagnosticsZipFilePath);
+
+            // upload the json metadata file
+            executionContext.Debug("Uploading diagnostic metadata file.");
+            string metadataFileName = $"diagnostics-{buildName}-{phaseName}.json";
+            string metadataFilePath = Path.Combine(supportFilesFolder, metadataFileName);
+            string phaseResult = GetTaskResultAsString(executionContext.Result);
+            
+            using (StreamWriter writer = File.CreateText(metadataFilePath)) 
+            {
+                writer.Write(JsonUtility.ToString(new DiagnosticLogMetadata(agentName, agentId.ToString(), poolId, phaseName, diagnosticsZipFileName, phaseResult)));
+            }
+
+            // CoreAttachmentType.DiagnosticLog
+            executionContext.QueueAttachFile(type: "DistributedTask.Core.DiagnosticLog", name: metadataFileName, filePath: metadataFilePath);
+
+            // CoreAttachmentType.DiagnosticLog
+            executionContext.QueueAttachFile(type: "DistributedTask.Core.DiagnosticLog", name: diagnosticsZipFileName, filePath: diagnosticsZipFilePath);
+
+            // Delete support folder
+            // TODO: We can't delete here. The file upload is queued so they will be gone when its time to upload. Will normal cleanup take care of it?
```

Comment:
you don't have to, since you put everything in agent temp, agent will clean them up.

Reply:
Removed comment.